### PR TITLE
Add an editSign() method to Player

### PR DESCRIPTION
**Pull request by Voltasalt**
This PR adds a method in Player that opens the sign editing dialog on the client. This can be used for simple sign editing plugins (similar to the one for Bukkit), getting input from the client, etc.

I'll implement it in SpongeCommon after this is merged.

### CODE REVIEW

**Comment by simon816** [comment]
Possibly have this method on [`Viewer`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/effect/Viewer.java) instead?

File: src/main/java/org/spongepowered/api/entity/player/Player.java [hunk 1]
```diff
@@ -24,12 +24,14 @@
...
 package org.spongepowered.api.entity.player;
 
+import org.spongepowered.api.block.tile.Sign;
 import org.spongepowered.api.data.manipulators.DisplayNameData;
 import org.spongepowered.api.data.manipulators.entities.GameModeData;
 import org.spongepowered.api.data.manipulators.entities.JoinData;
 import org.spongepowered.api.effect.Viewer;
 import org.spongepowered.api.entity.living.Human;
 import org.spongepowered.api.entity.player.tab.TabList;
+import org.spongepowered.api.event.block.tile.SignChangeEvent;
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/entity/player/Player.java]: Unused import

**Comment by Mumfrey** [comment]
@simon816 it doesn't really make sense on `Viewer`, both for the context of the interface itself and also the fact that only players can view GUIs

**Comment by ST-DDT** [comment]
Nice, could you clarify in the javadocs, that the method does **not** block the call flow/wait until the player has finished editing the sign.

**Comment by Zidane** [comment]
@Voltasalt 

Blast, I was gonna add this. A very cool feature that, imo, should have been allowed in Vanilla from the get go (edit a sign already placed).

**Comment by Voltasalt** [comment]
Would it be absolutely crazy to have some sort of callback parameter that'll get invoked once the player finishes editing? This would also mean that we could have the parameter be a Location, which could just be -1,-1,-1, and the callback would still be invoked with the lines even though the sign doesn't actually exist? This could be *really* useful for user input.

**Comment by Voltasalt** [comment]
To clarify, you would be able to do this:

player.editSign(new Location(-1, -1, -1), callbackThing), or even forego the location parameter entirely in an overload and *just* invoke the callback (without any actual sign existing)

**Comment by ST-DDT** [comment]
Maybe the method should contain `SignData` or something that contains the Text lines in that case
- With existing sign
  -  `editSign(Sign sign)`
- Without existing sign
  - `editSign(SignData data, Callback function)` 
  - `editSign(Text[] lines, Callback function)`

The `SignData` version has the benefit, that it can automatically contain new attributes for signs, because they will also be added to `SignData` and we don't have to adjust the method signature.

I'm not sure how the server should/can handle the callback though.

**Comment by boformer** [comment]
Why not just a SignEditEvent?

**Comment by ST-DDT** [comment]
@boformer  That is possible as well, 
`editSign(Text[] lines)` -> [SignChangeEvent](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/event/block/tile/SignChangeEvent.java)
but I would like to avoid the confusion whether the edited sign is virtual or not.

**Comment by simon816** [comment]
The sign would have to be in the world though, which is why the tile entity is the obvious choice. The client sends a `BlockPos` for the sign when updating it. If the sign is completely virtual (in the case of raw `SignData`), the packet `C12PacketUpdateSign` would need to be intercepted to see whether the reported sign change was indeed a real sign or a virtual one. This is kind of a hack so I'm not sure whether it should be possible to do this without the tile entity actually existing.

**Comment by Voltasalt** [comment]
The client accepts a packet with a bogus block position just fine:

        playerEntity.playerNetServerHandler.sendPacket(new S36PacketSignEditorOpen(new BlockPos(-1, -1, -1)));

However, to pre-fill the sign we'd have to place a sign somewhere in the world, send this packet, then remove the sign -- unless there's a better way?

**Comment by ryantheleach** [comment]
If you are going to pick magic values, you may as well ensure they are out of bounds of the vanilla world size as long as it doesn't cause problems.

**Comment by Voltasalt** [comment]
@ryantheleach Hence why I said "negative". Nothing can exist under y=0

**Comment by ryantheleach** [comment]
Yeah makes sense I guess as long as it doesn't cause problems anywhere. When you said negative I was thinking of x and z only, my bad.

**Comment by Voltasalt** [comment]
I'm not sure whether it would be best to use a Sign, a SignData or a Location.

**Comment by gabizou** [comment]
@Voltasalt I'd rather the use of `SignData` since we can't really fake a `Sign` TileEntity. I was also curious whether we could include some other things, such as a fake Villager Trade Window, or fake crafting, etc.

File: src/main/java/org/spongepowered/api/entity/player/Player.java [hunk 2]
```diff
@@ -134,4 +181,7 @@
      */
     GameModeData getGameModeData();
 
+    interface SignEditCallback {
+        void finishEdit(Player player, Text[] lines);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/player/Player.java]: Why not use SignData here as well?

File: src/main/java/org/spongepowered/api/entity/player/Player.java [hunk 3]
```diff
@@ -94,6 +96,51 @@
...
+     *
+     * <p>Any pre-existing lines on the {@link Sign} given will be filled in to the
+     * dialog, and the {@link Sign} will be edited with the contents of the dialog
+     * after the "Done" button is pressed by the player.</p>
+     *
+     * <p>This method does *not* block until the player has finished editing</p>
+     *
+     * @param sign The sign to edit
+     */
+    void editSign(SignData sign);
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/player/Player.java]: `SignData` does not contain a link to a real `Sign` TileEntity so i guess this should be just a `Sign` or a `Location`.

**Comment by Voltasalt** [comment]
@gabizou I'm not sure if any of the other types *can* be faked. If they can't, then they obviously can't be in the API :)

**Comment by SonarBeserk** [comment]
It seems like it would be better to have something like openSignGUI() or something. Maybe an outright method to take gui instances. Essentially this feels way too conditionally needed for my tastes, could use something more generic that provides the same features.

**Comment by ST-DDT** [comment]
You are right.
IMO we should create a new interface GUIConnected for Signs, NameTags, Crafting/Trading and other Inventories. That way we can open every kind of gui with a single method.
`openGUI(GUIConnected)` and `closeGUI()`.
Although GUIConnected sounds a little bit weird.

**Comment by ryantheleach** [comment]
NameTags..? You mean anvils right?

**Comment by ST-DDT** [comment]
@ryantheleach Oh yes that's true. I forgot that it has to be named on an anvil before it is used.
I hope my point is still clear.

**Comment by Aaron1011** [comment]
@SonarBeserk: I agree. There's currently no way to open guis, or determine how players interact with them. This is going to be a requirement for many bukkit plugins.

**Comment by Zidane** [comment]
This needs to be made generic...a way of opening not just one GUI but others.

I recommend re-thinking this and focus not on signs but guis as a whole.

**Pull request closed without merging**